mcomix.1
.TH MComix 1 "January 2016"

.SH NAME
.B MComix
\- GTK Comic Book Viewer

.SH SYNOPSIS
.B mcomix
.RI [ OPTION... "] [" PATH... ]

.SH DESCRIPTION
.B MComix
is a user-friendly, customizable image viewer. It is specifically designed to
handle comic books (both Western comics and manga) and supports a variety of
container formats (including CBR, CBZ, CB7, CBT, LHA and PDF).

.SH OPTIONS
.TP
.BR \-h , " \-\-help"
Display the help text and exit.
.TP
.BR \-s , " \-\-slideshow"
Start the application in slideshow mode.
.TP
.BR \-l , " \-\-library"
Show the library on startup.
.TP
.BR \-v , " \-\-version"
Show the version number and exit.
.PP
.B Viewing modes
.TP
.BR \-f , " \-\-fullscreen"
Start the application in fullscreen mode.
.TP
.BR \-m , " \-\-manga"
Start the application in manga mode.
.TP
.BR \-d , " \-\-double\-page"
Start the application in double page mode.
.PP
.B Zoom modes
.TP
.BR \-b , " \-\-zoom\-best"
Start the application with zoom set to best fit mode.
.TP
.BR \-w , " \-\-zoom\-width"
Start the application with zoom set to fit width.
.TP
.BR \-h , " \-\-zoom\-height"
Start the application with zoom set to fit height.
.PP
The zoom modes are mutually exclusive, and if more than one is specified,
the first will be used.

.SH ARGUMENTS
.TP
.I PATH
The path to one or more files or directories to be opened.

.SH DEPENDENCIES
.SS Required:
.B Python
3.3 or later
.PP
.B PyGObject
3.24 or later, with
.B GTK+ 3
3.22 or later.
.PP
.B PIL
(Python Imaging Library) 1.1.5 or later. Alternatively,
.B Pillow
(a fork of PIL) can be used.
.PP
.BR Distribute " or " Setuptools .
Distribute supersedes Setuptools on most distributions.
.SS Recommended:
.BR unrar ", " rar " or " libunrar
to extract RAR archives.
.PP
.B p7zip
to extract 7Z archives. Note that p7zip might be able to extract RAR archives
as well, but this might require additional software (for example,
.B p7zip-rar
on Debian-like systems), and it might fail to open certain RAR archives,
especially newer ones.
.PP
.B lha
to extract LHA archives.
.PP
.B mupdf
for PDF support.
.PP
.B pysqlite
(if not already built into Python).

.SH BUGS
.PD 0
Please report any bugs to the MComix bug tracker at
.PP
.I http://sourceforge.net/projects/mcomix

.SH CONTACT
.PD 0
The MComix home page is
.PP
.I http://mcomix.sourceforge.net


ChangeLog
MComix 1.2.1
Release date: 2016-02-12

    ENVIRONMENT/LOCALE/TRANSLATIONS:

    - Fixed a bug that made it impossible to open a book with MComix directly if
      the path contains spaces (Windows only)

    - Updated libraries for the Windows distribution: UnRAR DLL

MComix 1.2
Release date: 30.01.2016

    GUI/MAIN:

    - If metadata-based rotation is enabled, PNG files will be automatically
      rotated as well.

    - Double page mode respects Exif rotation now.

    - Some transformation issues have been fixed. In double page mode, all
      transformations are applied to the union of both pages. Also, reflection
      is performed first, followed by rotation.

    - Some OSD issues have been fixed.

    - When flipping pages, the content of the viewport does not appear
      somewhere else first anymore.

    - The default scaling quality is now "Bilinear".

    GUI/THUMBNAILER:

    - The thumbnailer now displays page numbers in a reasonable color
      appropriate for the respective background color.

    - The thumbnailer uses the same size for all thumbnails now. If thumbnails
      need to be rescaled, it is done using linear interpolation.

    - When using the keyboard, the thumbnailer now tries to keep the currently
      selected page in the upper half of its area.

    - The size of the thumbnailer is calculated more reasonably now.

    - Fixed a bug that could lead to crashes if the thumbnailer uses a dynamic
      background color.

    - The "missing image" icon appears in its original size in the thumbnailer.

    - Fix race condition that could lead to thumbnails being rendered with
      different sizes.

    - Added a workaround for a bug in gdk-pixbuf that could prevent thumbnails
      of animated GIF images from being rendered properly. For details, see
      https://bugzilla.gnome.org/show_bug.cgi?id=735422

    - Re-enabled double buffering for the thumbnailer.

    GUI/LIBRARY:

    - Some encoding issues with the library have been fixed.

    - The cover display in the library has been fixed.

    - The book area uses a tighter layout.

    - Various other issues with the library have been fixed.

    GUI/EDITARCHIVE:

    - Applying changes in the "Edit archive" dialog could raise an exception
      under certain circumstances. This has been fixed.

    - Fixed a bug that prevented MComix from shutting down properly if an
      archive with no images in it has been opened or the "Edit archive" dialog
      has been used.

    - Some issues with displaying thumbnails in the "Edit archive" dialog have
      been fixed.

    GUI/WM:

    - Fullscreen handling has been improved.

    - The Preferences dialog is not modal anymore.

    - When clicking on the thumbnailer while the main window is unfocused, the
      window should be focused only without switching to another page. This has
      been fixed so it works properly now.

    - Modal dialogs do not immediately hide the mouse cursor in the main area
      anymore.

    - Fix various minor window manager interaction issues.

    - The default window width is now 640 pixels.

    GUI/MISC:

    - You can select the text in the Properties dialog now.

    - The "Continue reading" dialog defaults to "Yes" now.

    - Dialogs refresh their respective contents whenever you switch to another
      page or book.

    - The password dialog now displays the path of the archive.

    - Overall widget handling has been fixed and improved. This also eliminates
      some GTK warnings.

    - Fixed a lot of issues with empty directories and empty archives.

    - File name filters and supported formats handling have been improved.

    - Various other issues with the Preferences dialog and the Properties
      dialog have been fixed.

    - Recently opened PDF files are now listed in the "Recent Files" menu.

    ENVIRONMENT/LOCALE/TRANSLATIONS:

    - The list of supported image formats is now determined dynamically,
      depending on the underlying libraries. This might implicitly add support
      for image formats such as WebP.

    - Due to a bug, PIL (or Pillow) was preferred over GdkPixbuf on Windows
      in earlier versions. Now, GdkPixbuf will be preferred on Windows as well.

    - MComix uses czipfile when available to speed up extraction of
      encrypted zip files.

    - Zombie processes will be removed if possible.

    - Some issues related to child processes have been fixed. Unnecessary
      console windows should not appear anymore.

    - File descriptors will be properly closed when possible. This fixes an
      issue especially on Windows where files used to stay "locked".

    - Searching for external tools (e.g. MuPDF) is performed more properly now.

    - Temporary directories will be created only when necessary and will be
      deleted as soon as the corresponding book has been closed.

    - The shebangs now ask for python2 instead of just python so we do not
      accidentally run Python 3.

    - comicthumb has been rewritten to make it consistent with MComix.

    - Some locale issues have been fixed.

    - The French translation has been updated.

    - The Russian translation has been updated (by Ulyanich Michael).

    - The Korean translation has been updated (by Gyeongmin Bak).

    - Fixed PDF support with newer versions of MuPDF (1.7 and 1.8).

    - Better support for using the 7z executable: encrypted files are now
      supported (including encrypted header support, and for all supported
      formats: 7z, RAR and ZIP).

    - Fixed an issue with unrar.dll that could lead to crashes if 7z is also
      present.

    - Improved detection of available RAR extractors. (unrar-free is currently
      incompatible with MComix and will be ignored.)

    - Fixed support for LHA archives (they were always marked as empty).

    - Fixed support for tar.xz archives (they were always marked as empty).

    - Updated libraries for the Windows distribution: Pillow 3.1.0 and

src/mcomix.py
#!/usr/bin/env python

"""MComix - GTK Comic Book Viewer
"""

# -------------------------------------------------------------------------
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# -------------------------------------------------------------------------

import os
import sys
import getopt
import gettext

_ = gettext.gettext

#Check for PyGTK and PIL dependencies.
try:
    import pygtk
    pygtk.require('2.0')

    import gtk
    assert gtk.gtk_version >= (2, 12, 0)
    assert gtk.pygtk_version >= (2, 12, 0)

    import gobject
    gobject.threads_init()

except AssertionError:
    print _("You don't have the required versions of GTK+ and/or PyGTK"),
    print _('installed.')
    print _('Installed GTK+ version is: ') + \
        '.'.join([str(n) for n in gtk.gtk_version])
    print _('Required GTK+ version is: 2.12.0 or higher\n')
    print _('Installed PyGTK version is: ') + \
        '.'.join([str(n) for n in gtk.pygtk_version])
    print _('Required PyGTK version is: 2.12.0 or higher')

    sys.exit(1)

except ImportError:
    print _('PyGTK version 2.12.0 or higher is required to run MComix.')
    print _('No version of PyGTK was found on your system.')

    sys.exit(1)

try:
    import Image
    assert Image.VERSION >= '1.1.5'

except AssertionError:
    print _("You don't have the required version of the Python Imaging"),
    print _('Library (PIL) installed.')
    print _('Installed PIL version is: %s') % Image.VERSION
    print _('Required PIL version is: 1.1.5 or higher')
    sys.exit(1)

except ImportError:
    print _('Python Imaging Library (PIL) 1.1.5 or higher is required.')
    print _('No version of the Python Imaging Library was found on your'),
    print _('system.')

    sys.exit(1)

import constants
import deprecated
import image_tools
import locale
import main
import icons
import preferences
import portability

def print_help():
    """Print the command-line help text and exit."""
    print _('Usage:')
    print '  mcomix ' + _('[OPTION...] [PATH]')
    print _('\nView images and comic book archives.\n')
    print _('Options:')
    print _('  -h, --help              Show this help and exit.')
    print _('  -f, --fullscreen        Start the application in fullscreen mode.')
    print _('  -l, --library           Show the library on startup.')

    sys.exit(1)


def run():
    """Run the program."""
    # Use gettext translations as found in the source dir, otherwise based on
    # the install path.
    exec_path = os.path.abspath(main.__file__)
    base_dir = os.path.dirname(os.path.dirname(exec_path))

    if os.path.isdir(os.path.join(base_dir, 'messages')):
        gettext.install('mcomix', os.path.join(base_dir, 'messages'),
            unicode=True)

    else:
        gettext.install('mcomix', os.path.join(base_dir, 'share/locale'),
            unicode=True)

    fullscreen = False
    show_library = False
    open_path = None
    open_page = 1

    try:
        argv = portability.get_commandline_args()
        opts, args = getopt.gnu_getopt(argv[1:], 'fhld',
            ['fullscreen', 'help', 'library'])

    except getopt.GetoptError:
        print_help()

    for opt, value in opts:

        if opt in ('-h', '--help'):
            print_help()

        elif opt in ('-f', '--fullscreen'):
            fullscreen = True

        elif opt in ('-l', '--library'):
            show_library = True

    if not os.path.exists(constants.DATA_DIR):
        os.makedirs(constants.DATA_DIR, 0700)

    if not os.path.exists(constants.CONFIG_DIR):
        os.makedirs(constants.CONFIG_DIR, 0700)

    deprecated.move_files_to_xdg_dirs()
    preferences.read_preferences_file()
    icons.load_icons()

    if len(args) >= 1:
        param_path = os.path.abspath(args[0])

        if os.path.isdir(param_path):
            dir_files = os.listdir(param_path)
            dir_files.sort(locale.strcoll)

            for filename in dir_files:
                full_path = os.path.join(param_path, filename)

                if image_tools.is_image_file(full_path):
                    open_path = full_path
                    break
        else:
            open_path = param_path

    elif preferences.prefs['auto load last file']:
        open_path = preferences.prefs['path to last file']
        open_page = preferences.prefs['page of last file']

    window = main.MainWindow(fullscreen=fullscreen, show_library=show_library,
        open_path=open_path, open_page=open_page)
    deprecated.check_for_deprecated_files(window)

    try:
        gtk.main()
    except KeyboardInterrupt: # Will not always work because of threading.
        window.save_and_terminate_program()

if __name__ == '__main__':
    run()

# vim: expandtab:sw=4:ts=4
